### DISCUSSION TITLE
encoder value is unstable

### DISCUSSION THREAD

**Discussion by yuuahmad** [grblHAL/core#444] [General]
The encoder that I use on my CNC lathe spindle cannot display a stable value.

I also did simple testing on my controller board by making a simple encoder simulation using an Arduino nano. here is a simple program

```
#include <Arduino.h>
const int pulsePin = 2;  
const int indexPin = 3; 
const int resolusiEncoder = 2500;
const int delayPulse = 100;
int pulseCount = 0; 

void setup()
{
  pinMode(pulsePin, OUTPUT);
  pinMode(indexPin, OUTPUT);
}

void loop()
{
  digitalWrite(pulsePin, HIGH);
  delayMicroseconds(delayPulse); 
  pulseCount++;
  if (pulseCount == resolusiEncoder)
  {
    digitalWrite(indexPin, HIGH);
    delayMicroseconds(delayPulse); 
    pulseCount = 0;
  }
  digitalWrite(pulsePin, LOW);
  digitalWrite(indexPin, LOW);
  delayMicroseconds(delayPulse);
}
```
after that I saw the encoder value that appeared in the verbose output. this is the result.
![Cuplikan layar 2024-02-14 162450](https://github.com/grblHAL/core/assets/22028802/5d4237d7-831e-4190-aa28-6be7edc26457)

Is this unstable value caused by firmware that has not been optimized?

**Comment by terjeio** [comment]
> Is this unstable value caused by firmware that has not been optimized?

It is not optimized for showing a stable RPM - I do not think that is really needed. What is important is that the reported angular position is accurate. Both uses data from the encoder.

Currently RPM is calculated from the time elapsed for a few number of pulses from the encoder, this could be increased - or a running average calculated. BTW reasons for the instability is likely both interrupt latency and how many pulses are used for the calulation. And one source could be the pulses coming from the Arduino not beeing 100% stable, again due to interrupt latency?

BTW which driver are you using and what is your PPR setting?

**Comment by yuuahmad** [comment]
here you go
my driver is STM32F446RE and the board is protoneer v3 with few modifications, and my ppr is 1024
```
>>> $i+
[VER:1.1f.20240127:]
[OPT:VNZH0SL,35,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,HOME,NOPROBE,ES,LATHE,SS,SED]
[FIRMWARE:grblHAL]
[SIGNALS:HSEFM]
[NVS STORAGE:*FLASH]
[FREE MEMORY:111K]
[DRIVER:STM32F446]
[DRIVER VERSION:240125]
[BOARD:Protoneer v3]
[AUX IO:0,2,0,0]
ok
>>> $$
$0 = 4.0    (Step pulse time, microseconds)
$1 = 25    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 0    (Step direction invert, mask)
$4 = 7    (Invert step enable pin, boolean)
$5 = 7    (Invert limit pins, boolean)
$9 = 3   
$10 = 511    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.001    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$14 = 6   
$15 = 0   
$16 = 0   
$17 = 0   
$18 = 0   
$20 = 1    (Soft limits enable, boolean)
$21 = 0    (Hard limits enable, boolean)
$22 = 11    (Homing cycle enable, boolean)
$23 = 0    (Homing direction invert, mask)
$24 = 100.0    (Homing locate feed rate, mm/min)
$25 = 1000.0    (Homing search seek rate, mm/min)
$26 = 250    (Homing switch debounce delay, milliseconds)
$27 = 4.000    (Homing switch pull-off distance, millimeters)
$28 = 0.100   
$29 = 0.0   
$30 = 4402.600    (Maximum spindle speed, RPM)
$31 = 5.000    (Minimum spindle speed, RPM)
$32 = 2    (Laser-mode enable, boolean)
$33 = 5000.0   
$34 = 0.0   
$35 = 0.0   
$36 = 100.0   
$37 = 0   
$38 = 1024   
$39 = 1   
$40 = 0   
$43 = 1   
$44 = 1   
$45 = 4   
$46 = 0   
$62 = 0   
$63 = 3   
$64 = 0   
$66 = ,,   
$67 = ,,   
$68 = ,,   
$69 = ,,   
$90 = 0.000   
$91 = 0.000   
$92 = 0.000   
$95 = 0.000   
$100 = 500.00000    (X-axis travel resolution, step/mm)
$101 = 500.00000    (Y-axis travel resolution, step/mm)
$102 = 500.00000    (Z-axis travel resolution, step/mm)
$110 = 8000.000    (X-axis maximum rate, mm/min)
$111 = 8000.000    (Y-axis maximum rate, mm/min)
$112 = 8000.000    (Z-axis maximum rate, mm/min)
$120 = 500.000    (X-axis acceleration, mm/sec^2)
$121 = 500.000    (Y-axis acceleration, mm/sec^2)
$122 = 500.000    (Z-axis acceleration, mm/sec^2)
$130 = 300.000    (X-axis maximum travel, millimeters)
$131 = 200.000    (Y-axis maximum travel, millimeters)
$132 = 200.000    (Z-axis maximum travel, millimeters)
$340 = 50.0   
$341 = 0   
$342 = 30.0   
$343 = 25.0   
$344 = 200.0   
$345 = 200.0   
$346 = 1   
$370 = 0   
$372 = 0   
$384 = 0   
$398 = 35   
$481 = 0   
$484 = 1   
$486 = 0   
ok
>>> $pins
[PIN:PA0,Emergency stop]
[PIN:PA1,Feed hold]
[PIN:PA4,Cycle start]
[PIN:PC7,X limit min]
[PIN:PB6,Y limit min]
[PIN:PA6,Z limit min]
[PIN:PC9,Spindle index]
[PIN:PA11,Aux input 0,Motor warning]
[PIN:PA12,Aux input 1,Motor fault]
[PIN:PA10,X step]
[PIN:PB3,Y step]
[PIN:PB5,Z step]
[PIN:PB4,X dir]
[PIN:PB10,Y dir]
[PIN:PA8,Z dir]
[PIN:PA9,Steppers enable]
[PIN:PB7,Spindle on]
[PIN:PA5,Spindle direction]
[PIN:PB0,Flood]
[PIN:PB15,Aux out 0,P0]
[PIN:PC12,Aux out 1,P1]
[PIN:PA3,RX,Primary UART]
[PIN:PA2,TX,Primary UART]
[PIN:PA7,Spindle PWM]
[PIN:PD2,Spindle pulse]
ok
```

**Comment by terjeio** [comment]
With that configuration I time 32 encoder pulses for calculating the RPM. When I feed the pulse input from a PWM generated signal it is generally very stable - displaying most RPMs the same for minutes on end (or perhaps as for long as I feed the pulse input),

So most likely most of the instability is due to jitter in the input. I could add a running average to smooth out the readings, but why would I do that? Do you have a technical reason for it?
Again, for [spindle synced motion](https://github.com/grblHAL/STM32F4xx/blob/019dc048242a669da8866ff7c01e21e60bfecac7/Src/driver.c#L942) the [angular position](https://github.com/grblHAL/STM32F4xx/blob/019dc048242a669da8866ff7c01e21e60bfecac7/Src/driver.c#L1599-L1647) is what is important and that cannot be smoothed out.
